[LiveMetrics] report process metrics `CPU Total` and `Committed Memory`

TimothyMothra opened the pull request:
### Changes
- LiveMetrics will now send Process metrics.
  - Memory: [Process.PrivateMemorySize64](https://learn.microsoft.com/dotnet/api/system.diagnostics.process.privatememorysize64)
  - CPU: [Process.TotalProcessorTime](https://learn.microsoft.com/dotnet/api/system.diagnostics.process.totalprocessortime)
     Measured as: `((change in ticks / period) / number of processors)`

## Testing
I've collected screenshots and detailed description for each counter.
For these metrics, I'm trying to match what Visual Studio's Debugger shows.

<details>
  <summary>Memory (click to expand)</summary>

To test I debugged an app in Visual Studio and compared to the Process Memory.

### Visual Studio Debugger
This shows roughly 50 MB of Process Memory
![Screenshot 2024-02-23 140056](https://github.com/Azure/azure-sdk-for-net/assets/28785781/87221d08-bd52-4440-a5a3-94e4f9a8f6f3)

### OTel uses: 
- WorkingSet64 
  - "Gets the amount of physical memory, in bytes, allocated for the associated process."
https://learn.microsoft.com/dotnet/api/system.diagnostics.process.workingset64
- [OTel repo link](https://github.com/open-telemetry/opentelemetry-dotnet-contrib/blob/de135415f5de21c0a565f3fb3b0570484e7a2bae/src/OpenTelemetry.Instrumentation.Process/ProcessMetrics.cs#L27)

This gives a value almost double of what Visual Studio Debugger is showing:
![Screenshot 2024-02-26 183506](https://github.com/Azure/azure-sdk-for-net/assets/28785781/b114338c-92ba-4fd1-9ff5-29fcfff8b535)


### AI SDK uses:
- PrivateMemorySize64 
  - "Gets the amount of private memory, in bytes, allocated for the associated process."
  https://learn.microsoft.com/dotnet/api/system.diagnostics.process.privatememorysize64
- [AI SDK repo link](https://github.com/microsoft/ApplicationInsights-dotnet/blob/bd8b0c25712e6f3c5275ae06b0f0e9331340fee4/WEB/Src/PerformanceCollector/Perf.Shared.NetStandard/Implementation/XPlatform/XPlatProcessMemoryPerformanceCounter.cs#L21)

This value closely matches Visual Studio's Diagnostic Tools:
![Screenshot 2024-02-23 140153](https://github.com/Azure/azure-sdk-for-net/assets/28785781/4a78c5dd-5e55-49b8-801b-3215143c2994)

</details>


<details>
  <summary>CPU (click to expand)</summary>


To test CPU I compute squareroots in a loop for 20 seconds.

### Visual Studio's Debugger:
Here, CPU jumps to 100% and stays high for a period before dropping back to 0%.
![Screenshot 2024-02-23 143342](https://github.com/Azure/azure-sdk-for-net/assets/28785781/b1ddbb42-47c9-4c30-80a5-bd3f34c55cd4)

### Otel uses 
- UserProcessorTime  
  "A TimeSpan that indicates the amount of time that the associated process has spent running code inside the application portion of the process (not inside the operating system core)."
  https://learn.microsoft.com/dotnet/api/system.diagnostics.process.userprocessortime
- and PrivilegedProcessorTime 
  "A TimeSpan that indicates the amount of time that the process has spent running code inside the operating system core."
  https://learn.microsoft.com/dotnet/api/system.diagnostics.process.privilegedprocessortime
- [OTel repo link](https://github.com/open-telemetry/opentelemetry-dotnet-contrib/blob/de135415f5de21c0a565f3fb3b0570484e7a2bae/src/OpenTelemetry.Instrumentation.Process/ProcessMetrics.cs#L48-L49)

Note that this grows and never resets to zero. That flat line at the end is when the compute finished.
![Screenshot 2024-02-23 143727](https://github.com/Azure/azure-sdk-for-net/assets/28785781/82e76f73-7512-47ce-8901-a9c907250519)



### AI SDK uses 
- TotalProcessorTime 
  "A TimeSpan that indicates the amount of time that the associated process has spent utilizing the CPU. This value is the sum of the UserProcessorTime and the PrivilegedProcessorTime."
  https://learn.microsoft.com/dotnet/api/system.diagnostics.process.totalprocessortime
- and measures the diff between two different collection timestamps, then divides by the number of processes
- [AI SDK repo link1](https://github.com/microsoft/ApplicationInsights-dotnet/blob/bd8b0c25712e6f3c5275ae06b0f0e9331340fee4/WEB/Src/PerformanceCollector/Perf.Shared.NetStandard/Implementation/XPlatform/XPlatProcessCPUPerformanceCounter.cs#L32-L59)
- [AI SDK repo link2](https://github.com/microsoft/ApplicationInsights-dotnet/blob/bd8b0c25712e6f3c5275ae06b0f0e9331340fee4/WEB/Src/PerformanceCollector/Perf.Shared.NetStandard/Implementation/XPlatform/XPlatProcessCPUPerformanceCounterNormalized.cs#L44)

TotalProcessorTime is defined as the sum of UserProcessorTime and PrivilegedProcessorTime. So this is the same metrics used by OTel. But AI SDK has a specific algorithm to calculate the difference.

Note that this more closely matches Visual Studio
![Screenshot 2024-02-23 144203](https://github.com/Azure/azure-sdk-for-net/assets/28785781/42f8f866-7ad0-471e-a656-9a1aef4fe184)








</details>

Commit:
implement process counters

Commit:
use Demo for testing. REVERT THIS BEFORE MERGING

azure-sdk commented:
**API change check**

API changes are not detected in this pull request.

Commit:
fix typo

rajkumar-rangaraj commented on sdk/monitor/Azure.Monitor.OpenTelemetry.LiveMetrics/src/Internals/Manager.Metrics.cs:
```diff
@@ -149,5 +164,70 @@ public MonitoringDataPoint GetDataPoint()
+        /// <summary>
+        /// Calcualte the CPU usage as the diff between two ticks divided by the period of time, and then divided by the number of processors.
+        /// </summary>
+        private bool TryCalculateCPUCounter(Process process, out double normalizedValue)
```
Why can't we use the same code as in process instrumentation - https://github.com/open-telemetry/opentelemetry-dotnet-contrib/blob/de135415f5de21c0a565f3fb3b0570484e7a2bae/src/OpenTelemetry.Instrumentation.Process/ProcessMetrics.cs#L41?

TimothyMothra replied:
The Process Instrumentation is not appropriate for our purposes.

1. The instrumentation is intending to record a "monotonically increasing value", that is a value that only increases.
  https://opentelemetry.io/docs/specs/otel/metrics/api/#asynchronous-counter The OpenTelemetry doc actually specifies CPU Time as an example. It's recording the ticks that the process is utilizing the CPU. This value will only ever increase.

    I tested reporting values to LiveMetrics in this way and that results in this chart: 
    Note that % Total increases past 100 during the stress test, but then flatlines when the stress test completes.
    ![image](https://github.com/Azure/azure-sdk-for-net/assets/28785781/60c344c0-1f8c-40a4-a368-e995d5dfeee8)


2. LiveMetrics is intending to show CPU utilization %. This value is expected to increase when under load, but then decrease back to zero. The Process counters only provide a "monotonically increasing" CPU tick. So we need to calculate the change over time.

    `((change in ticks / period) / number of processes)`
    
    ![image](https://github.com/Azure/azure-sdk-for-net/assets/28785781/3bce58c1-4727-445a-884c-57985ec918b3)


3. The OpenTelemetry instrumentation is reporting two counters separately (`process.UserProcessorTime` and `process.PrivilegedProcessorTime`). LiveMetrics doesn't need these separate values so I'm using `process.TotalProcessorTime` which is defined as [the sum of the UserProcessorTime and the PrivilegedProcessorTime.](
https://learn.microsoft.com/dotnet/api/system.diagnostics.process.totalprocessortime)


4. The OpenTelemetry instrumentation is recording `TotalSeconds`. I'm following AI SDK and using `Ticks`. I want to have the greatest possible precision in our calculations.

Commit:
changelog

cijothomas commented on sdk/monitor/Azure.Monitor.OpenTelemetry.LiveMetrics/CHANGELOG.md:
```diff
@@ -4,6 +4,9 @@
 
 ### Features Added
 
+- Report Performance Counters "CPU Total" and "Committed Memory".
```
do we want to use these names in changelog?

rajkumar-rangaraj replied:
Don't even call it as performance counters.

cijothomas commented on sdk/monitor/Azure.Monitor.OpenTelemetry.AspNetCore/tests/Azure.Monitor.OpenTelemetry.AspNetCore.Demo/Program.cs:
```diff
@@ -33,5 +38,52 @@
+    }
+
+    var timeStamp = DateTime.Now.AddSeconds(20);
+    while (DateTime.Now < timeStamp)
```
this will keep cpu busy for 20 secs..that was the intention, right?
If yes, modify the comment to say so!

TimothyMothra replied:
This will be reverted before merging.

rajkumar-rangaraj commented on sdk/monitor/Azure.Monitor.OpenTelemetry.AspNetCore/tests/Azure.Monitor.OpenTelemetry.AspNetCore.Demo/Program.cs:
```diff
@@ -33,5 +38,52 @@
     return $"Hello World! OpenTelemetry Trace: {Activity.Current?.Id}";
 });
 
+#if !NETFRAMEWORK
```
Do you want to keep the stress in the basic demo project?

TimothyMothra replied:
No. I mentioned in the PR description that this must be reverted before merging. 
I committed this so reviewers can see how I've been doing my testing. I also needed to save a copy of this because we've been iterating frequently and I needed to reuse the stress test.

rajkumar-rangaraj replied:
to avoid the confusion keep this changes locally in your dev env.

rajkumar-rangaraj commented on sdk/monitor/Azure.Monitor.OpenTelemetry.LiveMetrics/src/Internals/Manager.Metrics.cs:
```diff
@@ -149,5 +164,70 @@ public MonitoringDataPoint GetDataPoint()
+            var recentCollectedValue = this.cachedCollectedValue = process.TotalProcessorTime.Ticks;
+            var recentCollectedTime = this.cachedCollectedTime = DateTimeOffset.UtcNow;
+
+            var processorCount = Environment.ProcessorCount;
```
Move this out of this method, processor count won't vary.

rajkumar-rangaraj commented on sdk/monitor/Azure.Monitor.OpenTelemetry.LiveMetrics/src/Internals/Manager.Metrics.cs:
```diff
@@ -149,5 +164,70 @@ public MonitoringDataPoint GetDataPoint()
+                processorCount: processorCount,
+                normalizedValue: normalizedValue);
+            // TryCalculateCPUCounter period: 10313304 diff: 64062500 calculatedValue: 621.1636930318354 processorCount: 8 normalizedValue: 77.64546162897942
+            Debug.WriteLine($"{nameof(TryCalculateCPUCounter)} period: {period} diff: {diff} calculatedValue: {calculatedValue} processorCount: {processorCount} normalizedValue: {normalizedValue}");
```
I would prefer to remove this debug statement.

rajkumar-rangaraj commented:
perf counter is called out in code comment, title and variables. We are not getting data from perf counter. Use an alternate name instead.

Commit:
Revert "use Demo for testing. REVERT THIS BEFORE MERGING"

This reverts commit 38bf3f3a83eee996dc489643121bd2a472035cfb.

Commit:
cleanup

Commit:
update changelog

Commit:
refactor

Commit:
typo

Commit:
rename

Commit:
changelog

Commit:
rename

TimothyMothra commented:
@cijothomas @rajkumar-rangaraj 
I've addressed all feedback, please re-review.
I'll be out the rest of the week. If everything looks good feel free to merge. :)

rajkumar-rangaraj commented on sdk/monitor/Azure.Monitor.OpenTelemetry.LiveMetrics/src/Internals/Manager.Metrics.cs:
```diff
@@ -91,32 +92,47 @@ public MonitoringDataPoint GetDataPoint()
+        /// </remarks>
+        public IEnumerable<Models.MetricPoint> CollectProcessMetrics()
+        {
+            var process = Process.GetCurrentProcess();
```
We don't need to get the current process on every run. Missed in previous review.

rajkumar-rangaraj commented on sdk/monitor/Azure.Monitor.OpenTelemetry.LiveMetrics/src/Internals/Manager.Metrics.cs:
```diff
@@ -91,32 +92,47 @@ public MonitoringDataPoint GetDataPoint()
-        //    };
-        //}
+        /// <summary>
+        /// Collect metrics for the current process.
```
Better remove the these comment, until we have a clarity on process or processor metric.

rajkumar-rangaraj commented on sdk/monitor/Azure.Monitor.OpenTelemetry.LiveMetrics/CHANGELOG.md:
```diff
@@ -4,6 +4,9 @@
 
 ### Features Added
 
+- Report metrics "CPU Total" and "Committed Memory".
```
Just say report CPU and memory metrics.

@cijothomas Your thoughts?

reyang replied:
Can we make it precise and clear? This is one example of how to be precise and clear https://github.com/open-telemetry/opentelemetry-dotnet-contrib/tree/main/src/OpenTelemetry.Instrumentation.Process#processmemoryvirtual

cijothomas commented:
The screenshot in the description has one with CPU showing above 100%.. That should never happen, if you are sending normalized values.

reyang commented:
> The screenshot in the description has one with CPU showing above 100%.. That should never happen, if you are sending normalized values.

I suggest that we evaluate what .NET Extension did and align on the algorithm. At minimum, if we want to introduce our own implementation, I want the algorithm to be carefully designed, reviewed (and backed by Math) and documented. https://github.com/dotnet/extensions/blob/main/src/Libraries/Microsoft.Extensions.Diagnostics.ResourceMonitoring/Linux/LinuxUtilizationProvider.cs

cijothomas commented:
> > The screenshot in the description has one with CPU showing above 100%.. That should never happen, if you are sending normalized values.
> 
> I suggest that we evaluate what .NET Extension did and align on the algorithm. At minimum, if we want to introduce our own implementation, I want the algorithm to be carefully designed, reviewed (and backed by Math) and documented (e.g. this [PR](https://github.com/open-telemetry/opentelemetry-dotnet-contrib/pull/612) sets a good example). https://github.com/dotnet/extensions/blob/main/src/Libraries/Microsoft.Extensions.Diagnostics.ResourceMonitoring/Linux/LinuxUtilizationProvider.cs

Agree mostly. My thoughts are
1. AzMon distro or exporter should not be in the business of producing metrics. It should just bring in existing instrumentation libraries only. And so far, we seem to be sticking with this approach.
2. This particular PR is for the LiveMetrics feature. While it'd be nice to follow above here too, it may not be feasible/worth-the-effort. Considering LiveMetrics feature is not to store metrics anywhere permanent, the efforts to design and maintain an algorithm may not be worth it. (Its just for quick visual indication, not even queryable for later consumption, so carefully find the balance and avoid spending too much effort here)
3. We still need to do the basic accuracy part - i.e do not incorrectly tell what the CPU utilization is that of the machine, while actually showing just the single process cpu utilization. Do not state that the cpu utilization is normalized to take into account the number of cores, but send a non-normalized value. (these requires product changes too). (In fact, ApplicationInsights SDKs are lying in many cases - eg: In Linux or in AppService environmetns, app insights only sends process cpu and memory, but the UI shows it as Total Machine CPU and Memory, which is wrong, and that mistake should not be repeated here.) Again, this needs product change/support (And Mothra is already engaged with Product team to get the necessary changes)

rajkumar-rangaraj commented:
Providing a summary of the issue before we exploring into the specifics:

In the Application Insights SDK, for the Windows environment, CPU metrics are collected for the entire operating system. In contrast, for Azure App Service and other non-Windows environments, process-level CPU metrics are utilized. When it comes to memory metrics, Windows uses the performance counter \Memory\Committed Bytes, while other environments rely on process-level private bytes.

Languages other than .NET typically read the OS-level counter to send these metrics, leading to inconsistencies in the data sent across different environments and languages.

Mothra has initiated a collaboration with the services team to establish clear guidelines on what exactly should be sent. Our objective is to ensure that the same metrics are sent from all environments and languages.

The services team has requested additional time to formalize this process. As an interim solution, both service and SDK teams have agreed to use process-level metrics for both CPU and memory. This temporary solution will be implemented in beta versions, and we will not release a Release Candidate (RC) or General Availability (GA) version until the standardization is finalized. This approach allows us to continue our work on filters in the meantime.

> AzMon distro or exporter should not be in the business of producing metrics. It should just bring in existing instrumentation libraries only. And so far, we seem to be sticking with this approach.

This is correct, except for standard metrics. For live metrics, we don't want to take the dependency on the OpenTelemetry Metric SDK, as it will make the implementation very complex.

> This particular PR is for the LiveMetrics feature. While it'd be nice to follow above here too, it may not be feasible/worth-the-effort. 

Correct, it is not worth the effort. 

> Considering LiveMetrics feature is not to store metrics anywhere permanent, the efforts to design and maintain an algorithm may not be worth it. (Its just for quick visual indication, not even queryable for later consumption, so carefully find the balance and avoid spending too much effort here)

I have to agree with Cijo on this part. The algorithm used is very straightforward, and I don't see any issues with it. The [LinuxUtilizationProvider ](https://github.com/dotnet/extensions/blob/main/src/Libraries/Microsoft.Extensions.Diagnostics.ResourceMonitoring/Linux/LinuxUtilizationProvider.cs) uses the same approach..

> The screenshot in the description has one with CPU showing above 100%.. That should never happen, if you are sending normalized values.

This is not true. Mothra is trying to explain that using `Process.UserProcessorTime` plots an incorrect graph, and a small implementation is needed to solve it. He provided the accurate graph at the end of the description and has also validated the graph by onboarding the runtime metrics to the same app.

Finally, I need to acknowledge that we need to accurately capture the information in the changelogs.

cijothomas commented:
> The screenshot in the description has one with CPU showing above 100%.. That should never happen, if you are sending normalized values.
> 
> This is not true. Mothra is trying to explain that using Process.UserProcessorTime plots an incorrect graph, and a small implementation is needed to solve it. He provided the accurate graph at the end of the description and has also validated the graph by onboarding the runtime metrics to the same app.

Thanks! Yes I can see the description is pretty clear.

Commit:
pr feedback
